cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/queue/TimeEventQueueProcessor.java

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright © 2019 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.logging.pipeline.queue;
+
+import ch.qos.logback.classic.spi.ILoggingEvent;
+import co.cask.cdap.api.metrics.MetricsContext;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.logging.meta.Checkpoint;
+import co.cask.cdap.logging.pipeline.LogProcessorPipelineContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.function.Function;
+
+/**
+ * The {@link TimeEventQueue} processor to enqueue the log events to {@link TimeEventQueue}, and process them.
+ * @param <Offset> type of the offset
+ */
+public class TimeEventQueueProcessor<Offset extends Comparable<Offset>> {
+  private static final Logger LOG = LoggerFactory.getLogger(TimeEventQueueProcessor.class);
+  private static final double MIN_FREE_FACTOR = 0.5d;
+  private final TimeEventQueue<ILoggingEvent, Offset> eventQueue;
+  private final LogProcessorPipelineContext context;
+  private final MetricsContext metricsContext;
+  private final long maxBufferSize;
+  private final long eventDelayMillis;
+
+  /**
+   * Time event queue processor.
+   */
+  public TimeEventQueueProcessor(LogProcessorPipelineContext context, long maxBufferSize, long eventDelayMillis,
+                                 Iterable<Integer> partitions) {
+    this.context = context;
+    this.maxBufferSize = maxBufferSize;
+    this.eventDelayMillis = eventDelayMillis;
+    this.metricsContext = context;
+    this.eventQueue = new TimeEventQueue<>(partitions);
+  }
+
+  /**
+   * Processes events provided by event iterator for a given partition.
+   *
+   * @param partition log event partition
+   * @param eventIterator log events iterator
+   * @param eventConverter event converter to convert log event to {@code ProcessorEvent}
+   * @param <T> type of the event provided by the iterator
+   *
+   * @return processed event metadata
+   */
+  public <T> ProcessedEventMetadata<Offset> process(int partition, Iterator<T> eventIterator,
```

Comment:
This method should simply takes a `Iterator<ProcessorEvent<Offset>>` directly. The caller can do transformation in any way they like.